astropy/coordinates/sky_coordinate.py

```diff
@@ -900,9 +903,13 @@ def match_to_catalog_3d(self, catalogcoord, nthneighbor=1):
         """
         from .matching import match_coordinates_3d
 
+
         if (isinstance(catalogcoord, (SkyCoord, BaseCoordinateFrame))
                 and catalogcoord.has_data):
-            self_in_catalog_frame = self.transform_to(catalogcoord)
+            if self.is_equivalent_frame(catalogcoord):
+                self_in_catalog_frame = self
+            else:
+                self_in_catalog_frame = self.transform_to(catalogcoord)
```

Comment:
Now, looking at `matching,py`, I see that the first think `match_coordinates_sky` does is to transform the coordinate to the catalog frame: https://github.com/astropy/astropy/blob/master/astropy/coordinates/matching.py#L72
So, these lines should not be here in the first place and we should adjust the code in `matching.py`.

I don't really have time to delve into this now; shall I just undo this change and raise a new issue?